apm-commons/apm-datacarrier/src/main/java/org/apache/skywalking/apm/commons/datacarrier/common/AtomicRangeInteger.java

```diff
@@ -18,23 +18,32 @@
 
 
 package org.apache.skywalking.apm.commons.datacarrier.common;
-
 import java.io.Serializable;
 import java.util.concurrent.atomic.AtomicInteger;
 
 /**
- * Created by wusheng on 2016/10/25.
+ * Created by lkxiaolou on 2019/06/22.
  */
 public class AtomicRangeInteger extends Number implements Serializable {
     private static final long serialVersionUID = -4099792402691141643L;
+
+    /**
+     * please dont't change the left and right Placeholders
+     * to disable false-sharing for more performance
+     */
+    private int[] leftPlaceholders;
     private AtomicInteger value;
+    private int[] rightPlaceholders;
+
     private int startValue;
     private int endValue;
 
     public AtomicRangeInteger(int startValue, int maxValue) {
         this.value = new AtomicInteger(startValue);
         this.startValue = startValue;
         this.endValue = maxValue - 1;
+        leftPlaceholders = new int[15];
+        rightPlaceholders = new int[15];
```

Comment:
I don't find the place to use `leftPlaceholders` and `rightPlaceholders`. And why the length is 15?

Reply:
I think I need a little more explanation about this.

Reply:
sorry, I have make some tests to find my first version is right. so I'm roll back the code.
Use an array ensure the value load in CPU's cache and cached in one line without other variable.

We can learn the Disruptor how to do this in this article

[https://tech.meituan.com/2016/11/18/disruptor.html](https://tech.meituan.com/2016/11/18/disruptor.html)

The Unsafe is used in the Disruptor, and it worked well. So we don't worry about Unsafe will change in the future.

Reply:
> I don't find the place to use `leftPlaceholders` and `rightPlaceholders`. And why the length is 15?

It's indeed not used in the code because it's only declared to pad the cache line (64 bytes per line ), 15 * 4 = 60,  plus the value to occupy the entire cache line to dismiss the false-sharing, am I right ? @lkxiaolou but why do we need padding right? Did I miss anything ?

Reply:
> sorry, I have make some tests to find my first version is right. so I'm roll back the code.
> Use an array ensure the value load in CPU's cache and cached in one line without other variable.
> 
> We can learn the Disruptor how to do this in this article
> 
> https://tech.meituan.com/2016/11/18/disruptor.html
> 
> The Unsafe is used in the Disruptor, and it worked well. So we don't worry about Unsafe will change in the future.

@wu-sheng the unsafe is also used in netty too, I don't think it's a concern herr

Reply:
> > I don't find the place to use `leftPlaceholders` and `rightPlaceholders`. And why the length is 15?
> 
> It's indeed not used in the code because it's only declared to pad the cache line (64 bytes per line ), 15 * 4 = 60, plus the value to occupy the entire cache line to dismiss the false-sharing, am I right ? @lkxiaolou but why do we need padding right? Did I miss anything ?

The value may cached  anywhere in one line,.
 We must ensure one cache line cached the value and can not cache other variables.
 So suppose that cache the value in the end slot of the cache line, we need 15 left and 15 right.

Reply:
> > > I don't find the place to use `leftPlaceholders` and `rightPlaceholders`. And why the length is 15?
> > 
> > 
> > It's indeed not used in the code because it's only declared to pad the cache line (64 bytes per line ), 15 * 4 = 60, plus the value to occupy the entire cache line to dismiss the false-sharing, am I right ? @lkxiaolou but why do we need padding right? Did I miss anything ?
> 
> The value may cached anywhere in one line,.
> We must ensure one cache line cached the value and can not cache other variables.
> So suppose that cache the value in the end slot of the cache line or start slot of the cache line, we need 15 left and 15 right.

Got it, really interesting